pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go

```diff
@@ -652,39 +658,57 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
 		return b.checkpoint, nil
 	}
 
+	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
+
 	// Ensure the batcher is always reset, even on early error returns.
 	preFlushTime := timeutil.Now()
 	lrw.debug.RecordFlushStart(preFlushTime, int64(len(b.buffer.curKVBatch)))
 
+	workers := min(maxWriterWorkers, (len(b.buffer.curKVBatch)/batchSize)+1)
+	perWorker := (len(b.buffer.curKVBatch) / workers) + 1
 	// TODO: The batching here in production would need to be much
 	// smarter. Namely, we don't want to include updates to the
 	// same key in the same batch. Also, it's possible batching
 	// will make things much worse in practice.
-	flushByteSize := 0
-	batchStart := 0
-	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
-	batchEnd := min(batchStart+batchSize, len(b.buffer.curKVBatch))
-	for batchStart < len(b.buffer.curKVBatch) && batchEnd != 0 {
-		preBatchTime := timeutil.Now()
-		batchStats, err := lrw.bh.HandleBatch(ctx, b.buffer.curKVBatch[batchStart:batchEnd])
-		if err != nil {
-			// TODO(ssd): Handle errors. We should perhaps split the batch and retry a portion of the batch.
-			// If that fails, send the failed application to the dead-letter-queue.
-			return nil, err
+	var flushByteSize atomic.Int64
+
+	g := ctxgroup.WithContext(ctx)
+	for i := 0; i < workers; i++ {
+		bh := lrw.bh[i]
+		batchStart := i * perWorker
+		chunkEnd := batchStart + perWorker
+		if i == workers-1 {
+			chunkEnd = len(b.buffer.curKVBatch)
 		}
 
-		flushByteSize += batchStats.byteSize
-		batchStart = batchEnd
+		g.GoCtx(func(ctx context.Context) error {
```

Comment:
i wonder if we'll notice some sort of overhead around creating up to 32 goroutines every time we flush. Perhaps in the next iteration, we ought to keep the workers around across flushes.

Reply:
future work